Q24ECU/core/include/hal/hal_spi.h

```diff
@@ -0,0 +1,471 @@
+#define SPI4_PCLK_EN() (RCC->APB2ENR |= (1 << 13))
+
+#define SPI1_PCLK_DI() (RCC->APB2ENR &= ~(1 << 12))
+#define SPI2_PCLK_DI() (RCC->APB1ENR &= ~(1 << 14))
+#define SPI3_PCLK_DI() (RCC->APB1ENR &= ~(1 << 15))
+#define SPI4_PCLK_DI() (RCC->APB2ENR &= ~(1 << 13))
+
+// Register Reset - TODO: check if the right registers were used
+#define SPI1_REG_RESET()               \
+    do                                 \
+    {                                  \
+        (RCC->APB2RSTR |= (1 << 12));  \
+        (RCC->APB2RSTR &= ~(1 << 12)); \
+    } while (0)
+
+#define SPI2_REG_RESET()               \
+    do                                 \
+    {                                  \
+        (RCC->APB1RSTR |= (1 << 14));  \
+        (RCC->APB1RSTR &= ~(1 << 14)); \
+    } while (0)
+
+#define SPI3_REG_RESET()               \
+    do                                 \
+    {                                  \
+        (RCC->APB1RSTR |= (1 << 15));  \
+        (RCC->APB1RSTR &= ~(1 << 15)); \
+    } while (0)
+
+#define SPI4_REG_RESET()               \
+    do                                 \
+    {                                  \
+        (RCC->APB2RSTR |= (1 << 13));  \
+        (RCC->APB2RSTR &= ~(1 << 13)); \
+    } while (0)
+
+// SPI_DeviceMode
+#define SPI_DEVICE_MODE_SLAVE 0
+#define SPI_DEVICE_MODE_MASTER 1
+// SPI_BusConfig
+#define SPI_BUS_CONFIG_FD 1
+#define SPI_BUS_CONFIG_HD 2
+#define SPI_BUS_CONFIG_SIMPLEX_RXONLY 3
+// SPI_SclkSpeed
+#define SPI_SCLK_SPEED_DIV2 0
+#define SPI_SCLK_SPEED_DIV4 1
+#define SPI_SCLK_SPEED_DIV8 2
+#define SPI_SCLK_SPEED_DIV16 3
+#define SPI_SCLK_SPEED_DIV32 4
+#define SPI_SCLK_SPEED_DIV64 5
+#define SPI_SCLK_SPEED_DIV128 6
+#define SPI_SCLK_SPEED_DIV256 7
+// SPI_DFF
+#define SPI_DFF_8BITS 0
+#define SPI_DFF_16BITS 1
+// SPI_CPOL
+#define SPI_CPOL_LOW 0
+#define SPI_CPOL_HIGH 1
+// SPI_CPHA
+#define SPI_CPHA_LOW 0
+#define SPI_CPHA_HIGH 1
+// SPI_SSM
+#define SPI_SSM_DI 0
+#define SPI_SSM_EN 1
+
+// SPI Application States - used for interrupts mostly
+#define SPI_READY 0
+#define SPI_BUSY_IN_RX 1
+#define SPI_BUSY_IN_TX 2
+
+// Possible SPI application events
+#define SPI_EVENT_TX_CMPLT 1
+#define SPI_EVENT_RX_CMPLT 2
+#define SPI_EVENT_OVR_ERR 3
+#define SPI_EVENT_CRC_ERR 4
+
+// Configuration structure for SPIx peripheral
+typedef struct
+{
+    uint8_t SPI_DeviceMode;
+    uint8_t SPI_BusConfig;
+    uint8_t SPI_SclkSpeed;
+    uint8_t SPI_DFF;  // Data frame format
+    uint8_t SPI_CPOL; // Polarity
+    uint8_t SPI_CPHA; // Phase
+    uint8_t SPI_SSM;  // Slave Management (Software or Hardware)
+} SPI_Config_t;
+
+// Handle structure for SPIx peripheral
+typedef struct
+{
+    SPI_TypeDef *pSPIx; // Holds the base address of the SPI peripheral. Struct overlay for the SPI registers.
+    SPI_Config_t SPIConfig;
+    uint8_t *pTxBuffer;
+    uint8_t *pRxBuffer;
+    uint32_t TxLen;
+    uint32_t RxLen;
+    uint8_t TxState;
+    uint8_t RxState;
+} SPI_Handle_t;
```

Comment:
This should disappear when you move all of the interrupt code out of this file anyways

Reply:
This has SPI_TypeDef as one of its members. It's used for all of the interrupt functions because it has important fields like RxState etc. It's more similar to this struct in the uart interface: 
![image](https://github.com/qfsae/zenith/assets/75098435/85dc9882-89b2-4098-9fe9-cf9034b8687f)
It also has the config struct as one of its members, so the hal init function is still dependent on it. I will leave it in the HAL file for now and look at refactoring it in the interface PR

Reply:
Okok I see how this is being used....

Reply:
Tbh I really don't know what to do with this bc it shouldn't be here and if your using it like the one I dumped in the UART file it should be in the interface and it should use freertos streambuffers/semaphores.

Reply:
Fixed - When we do the interface PR, we will add freertos mechanisms like semaphores.